libopenage/gui/gui_textured_points_shader.h

```diff
@@ -0,0 +1,97 @@
+// Copyright 2016-2016 the openage authors. See copying.md for legal info.
+
+#pragma once
+
+#include <QSGSimpleMaterialShader>
+#include <QSGTextureProvider>
+
+namespace openage {
+namespace gui {
+
+struct Icon
+{
+	QQuickItem *marker_atlas;
+	float columns;
+	float spacing;
+};
+
+/**
+ * Draws ColoredTexturedPoint2D arrays.
+ *
+ * Uses icons found in the square-grid atlas by the ColoredTexturedPoint2D::cell_id.
+ * Applies color from ColoredTexturedPoint2D regardless of the color in the atlas.
+ */
+class TexturedPointsShader : public QSGSimpleMaterialShader<Icon>
+{
+	QSG_DECLARE_SIMPLE_SHADER(TexturedPointsShader, Icon)
+
+public:
+	TexturedPointsShader()
+		:
+		QSGSimpleMaterialShader<Icon>{},
+		id_tex{},
+		id_columns{},
+		id_spacing{} {
+	}
+
+	virtual const char* vertexShader() const override {
+		return "#version 120\n"
+			"attribute highp vec4 vertex;\n"
+			"attribute highp float cell_id;\n"
+			"attribute lowp vec4 color;\n"
+			"uniform highp mat4 qt_Matrix;\n"
+			"varying vec4 v_color;\n"
+			"varying float v_cell_id;\n"
+			"void main() {\n"
+			"    gl_Position = qt_Matrix * vertex;\n"
+			"    v_color = color;\n"
+			"    v_cell_id = cell_id;\n"
+			"}";
+	}
+
+	virtual const char* fragmentShader() const override {
+		return "#version 120\n"
+			"uniform lowp float qt_Opacity;\n"
+			"uniform lowp sampler2D tex;\n"
+			"uniform highp float columns;\n"
+			"uniform highp float spacing;\n"
+			"varying vec4 v_color;\n"
+			"varying float v_cell_id;\n"
+			"void main() {\n"
+			"    vec2 border = vec2(spacing);\n"
+			"    vec2 in_grid_coord = vec2(mod(v_cell_id, columns), floor(v_cell_id / columns)) + border;\n"
+			"    vec2 subtex_coord = (in_grid_coord + gl_PointCoord * (vec2(1.0) - border)) / columns;\n"
+			"    if (texture2D(tex, subtex_coord).a < 0.1)\n"
+			"        discard;\n"
+			"    else;\n"
+			"        gl_FragColor = vec4(v_color.rgb, v_color.a * texture2D(tex, subtex_coord).a) * qt_Opacity;\n"
+			"}";
```

Comment:
can we put the shaders to the others in `assets/shaders/`? Would be better to integrate that after my #723 because it brings all the file loading code.